File: frontend/src/microservices/flight/objects/FlightBookingWindowStatus.ts
// 本文件定义 flight 模块的 `FlightBookingWindowStatus`，作为状态枚举并提供 JSON 编解码。

export type FlightBookingWindowStatus = 'Available' | 'SurchargeRequired' | 'Expired'
export const flightBookingWindowStatusFromJson = (json: string): FlightBookingWindowStatus =>
  JSON.parse(json) as FlightBookingWindowStatus

export const flightBookingWindowStatusToJson = (value: FlightBookingWindowStatus): string =>
  JSON.stringify(value)

File: backend/src/main/scala/microservices/flight/objects/FlightBookingWindowStatus.scala
// FlightBookingWindowStatus is a backend-computed status derived from booking-window rules and timestamps.
// It is intentionally backend-only because the frontend should consume the computed outcome, not reimplement the timing and surcharge rules.
package com.typesafe.travel.flight.objects

import io.circe.{Decoder, Encoder}

final case class FlightBookingWindowStatus(value: String):
  override def toString: String = value

object FlightBookingWindowStatus:
  val Available = FlightBookingWindowStatus("Available")
  val SurchargeRequired = FlightBookingWindowStatus("SurchargeRequired")
  val Expired = FlightBookingWindowStatus("Expired")

  given sourceEncoder: Encoder[FlightBookingWindowStatus] = Encoder.encodeString.contramap(_.toString)
  given sourceDecoder: Decoder[FlightBookingWindowStatus] = Decoder.decodeString.map(fromText)

  def fromText(value: String): FlightBookingWindowStatus =
    value.trim match
      case "Available" => Available
      case "SurchargeRequired" => SurchargeRequired
      case "Expired" => Expired
      case other => FlightBookingWindowStatus(other)

